### PR TITLE
feat(planner): improve push down filter join

### PR DESCRIPTION
**Pull request by Dousir9**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

1. Introduce can_filter_null to convert more outer joins into inner joins
2. Improve InferFilterOptimizer to support predicate derivation of outer/semi/anti join, for example, for sql `select * from t1 left join t2 on t1.a = t2.a where t1.a > 0`, we can get new filter `t2.a > 0` and use it to filter out more hash join build side data in advance;

- Fixes #[Link the issue here]

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14872)

### CODE REVIEW

File: src/query/sql/src/executor/physical_plans/physical_hash_join.rs [hunk 1]
```diff
@@ -82,9 +82,9 @@ pub struct HashJoin {
     pub enable_bloom_runtime_filter: bool,
     // Under cluster, mark if the join is broadcast join.
     pub broadcast: bool,
-    // Original join type. Left/Right single join may be convert to inner join
-    // Record the original join type and do some special processing during runtime.
-    pub original_join_type: Option<JoinType>,
+    // When left/right single join converted to inner join, record the original join type
+    // and do some special processing during runtime.
+    pub single_to_inner: Option<JoinType>,
```
Comment (xudong963) [inline src/query/sql/src/executor/physical_plans/physical_hash_join.rs:87]: `single_to_inner` is more like a bool type which represents if single to inner join has happened.

Comment (Dousir9) [reply src/query/sql/src/executor/physical_plans/physical_hash_join.rs:87]: Yes, we also need to record the original join type.

File: tests/sqllogictests/suites/mode/standalone/explain/join.test [hunk 2]
```diff
@@ -825,19 +829,15 @@ EvalScalar
...
-    │       └── estimated rows: 0.00
+    ├── TableScan(Build)
+    │   ├── table: default.default.t2
+    │   ├── output columns: [wms_id (#4)]
+    │   ├── read rows: 0
+    │   ├── read bytes: 0
+    │   ├── partitions total: 0
+    │   ├── partitions scanned: 0
+    │   ├── push downs: [filters: [], limit: NONE]
+    │   └── estimated rows: 0.00
```
Comment (leiysky) [inline tests/sqllogictests/suites/mode/standalone/explain/join.test:840]: Seems the filter `substr(t2.wms_id, 1) = 'R006'` is failed to push down?

Comment (Dousir9) [reply tests/sqllogictests/suites/mode/standalone/explain/join.test]: Yes, fixed

**Pull request merged**